crates/sui-indexer/src/handlers/objects_snapshot_processor.rs

```diff
@@ -62,105 +92,220 @@ impl Default for SnapshotLagConfig {
+    async fn process_checkpoint(&self, checkpoint: CheckpointData) -> anyhow::Result<()> {
+        let checkpoint_sequence_number = checkpoint.checkpoint_summary.sequence_number;
+        // Index the object changes and send them to the committer.
+        let object_changes: TransactionObjectChangesToCommit =
+            CheckpointHandler::<S, T>::index_objects(
+                checkpoint,
+                &self.metrics,
+                self.package_resolver.clone(),
+            )
+            .await?;
+        self.indexed_obj_sender
+            .send(CheckpointObjectChanges {
+                checkpoint_sequence_number,
+                object_changes,
+            })
+            .await?;
+        Ok(())
+    }
+
+    fn preprocess_hook(&self, checkpoint: CheckpointData) -> anyhow::Result<()> {
+        let package_objects = CheckpointHandler::<S, T>::get_package_objects(&[checkpoint]);
+        self.package_buffer
+            .lock()
+            .unwrap()
+            .insert_packages(package_objects);
+        Ok(())
+    }
+}
+
+// Start both the ingestion pipeline and committer for objects snapshot table.
+pub async fn start_objects_snapshot_processor<S, T>(
+    store: S,
+    metrics: IndexerMetrics,
+    config: IndexerConfig,
+    snapshot_config: SnapshotLagConfig,
+    reader_options: ReaderOptions,
+    cancel: CancellationToken,
+) -> IndexerResult<()>
+where
+    S: IndexerStore + Clone + Sync + Send + 'static,
+    T: R2D2Connection + 'static,
+{
+    info!("Starting object snapshot processor...");
+
+    let watermark = store
+        .get_latest_object_snapshot_checkpoint_sequence_number()
+        .await
+        .expect("Failed to get latest snapshot checkpoint sequence number from DB")
+        .map(|seq| seq + 1)
+        .unwrap_or_default();
+    let download_queue_size = std::env::var("DOWNLOAD_QUEUE_SIZE")
+        .unwrap_or_else(|_| crate::indexer::DOWNLOAD_QUEUE_SIZE.to_string())
+        .parse::<usize>()
+        .expect("Invalid DOWNLOAD_QUEUE_SIZE");
+
+    // Set up exit sender and receiver. Sender signals the ingestion executor to exit when
+    // the cancel token is triggered.
+    let (exit_sender, exit_receiver) = oneshot::channel();
+    let cancel_clone = cancel.clone();
+    spawn_monitored_task!(async move {
+        cancel_clone.cancelled().await;
+        let _ = exit_sender.send(());
+    });
+
+    // Commit notifier to to signal the package buffer that a checkpoint has been committed.
+    let (commit_notifier, commit_receiver) = watch::channel(None);
+
+    let global_metrics = get_metrics().unwrap();
+    // Channel for actually communicating indexed object changes between the ingestion pipeline and the committer.
+    let (indexed_obj_sender, indexed_obj_receiver) = mysten_metrics::metered_channel::channel(
+        // TODO: placeholder for now
+        1800,
+        &global_metrics
+            .channel_inflight
+            .with_label_values(&["obj_indexing_for_snapshot"]),
+    );
```

Comment:
what happens when the channel is full? do we stall until the next 600 checkpoints are streamed?

Reply:
Yes we stall on the sender side until the reader pops something off of the channel.